examples/server/server.cpp

```diff
@@ -377,12 +406,22 @@ struct llama_server_context {
         return stop_pos;
     }
 
-    std::string doCompletion() {
-        const llama_token token = nextToken();
+    completion_string_output doCompletion() {
+        const completion_token_output token_with_probs = nextToken();
+        completion_string_output result;
 
-        const std::string token_text = token == -1 ? "" : llama_token_to_str(ctx, token);
+        const std::string token_text = token_with_probs.tok == -1 ? "" : llama_token_to_str(ctx, token_with_probs.tok);
+        result.tok_str = token_text;
         generated_text += token_text;
 
+        // iterate through token_with_probs.probs, if tok is valid, convert it to string and add to result.prob
+        for (const auto & prob : token_with_probs.probs) {
+            const std::string prob_text = prob.tok == -1 ? "" : llama_token_to_str(ctx, prob.tok);
```

Comment:
It is not checking for validity? It will crash the server when a partial UTF-8 character is generated?

Maybe it should happen after the multibyte stuff below?

Reply:
My bad. I didn't take care of partial character generation at all in my first commit. Just added support in the newest commit.

> It would only apply if the other probable tokens were.

I now bundle the completion_token_output objects when dealing with multibyte. ie. if the character is 2 tokens long, we will return an array of 2 completion_token_output objects.

> I would prefer if all the code for token probabilities would be behind a flag in the request so it doesn't add overhead when is not needed.

Agree. Implemented this in the newest commit.